### PR TITLE
Formula bar attempt 3

### PR DESCRIPTION
**Pull request by aarondr77**
# Description

Makes the formula bar work the same way as the cell editor!

Note: I'm logging the location of the cell editor by passing a param. Works fine, but I don't think we pass uneeded params elsewhere, so lmk if its okay. 

# Testing

I tested this on mac: [firefox, chrome] and windows: [brave, chrome], and it works as expected. Can you pick a different combo?

# Documentation

I added a tiny bit of documentation because there isn't a good place to add it right now. When we redo the formula writing documentation page, we should def highlight this.

### CODE REVIEW

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:25]: handeled -> handled

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (naterush) [reply mitosheet/src/components/endo/celleditor/CellEditor.tsx:50]: Ah, I see what you're doing. You're attempting to watch for changes on the id, and then running this effect when it is set and not undefined.

This only works currently due to luck, I think, and other pre-renders that are happening. The key is that setting a ref or changing a ref generally does not cause effects to rerun, or the react component to re-render generally. 

You're looking for a callback ref. Peep line 251 of the the Dropdown.tsx for an example of how to run a function after the ref is defined. 

Also, sorry if you took this code from somewhere else and I wrote it. I'm also learning about this stuff in real time, so parts of our codebase do this just like you're doing it here - even though it's fragile and pretty much bad!

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:430]: Why would it be undefined?

File: mitosheet/src/components/endo/celleditor/FloatingCellEditor.tsx
Comment (naterush) [reply mitosheet/src/components/endo/celleditor/FloatingCellEditor.tsx:121]: We previously didn't make it do this, which I think is the right call and fine. Let's just update the comment then perhaps.

**Review by naterush — CHANGES_REQUESTED** [review]
Working great. Some minor cleanup, and then ping me and I'll get it merged!

**Comment by naterush** [comment]
Screen Shot 2022-04-19 at 5 02 23 PM

A few styling / polish questions that I think will take this feature up another notch on top of the improvement it already is: 
1. Does the harsh purple feel really out of place up top? I think bumping it to 1px and a higher purple looks a lot better. let me know what you think. 
2. Do we want the suggestion box to get a bit bigger up top? I feel like we can utilize a lot more space up here. I don't know if it needs to get massive, but perhaps like twice as big? Your call here. 
3. There's whitespace between the cell editor and the rest of the formula bar that looks kinda terrible. Like if you start editing, there's the purple border, white space, and then the bottom border of the toolbar. Can we get rid of that white space?

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (aarondr77) [reply mitosheet/src/components/endo/celleditor/CellEditor.tsx]: I'm currently using this suggestion, but there still is some race condition with the focus. So I have a really tiny timeout. I can't figure out where the race condition is though. If you want to give it a try feel free to.

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:53]: I think we can just call unsavedInputAnchor.focus - but this looks much better!

**Comment by naterush** [comment]
@aarondr77 ok update I found one bug in this in final test:
1. Create a new column
2. Click on a column header of the new column in the sheet
3. Double click on the formula bar. 
4. Observe that the formula for for the column (=0) gets submitted as the column header

Pretty much, we don't support header editing in the formula bar. Let's fix this up before merging, either:
1. Making it so you can edit column headers in the formula bar, and it just works.
2. We just turn off formula bar editing for headers (this is worse). 

I think we should do (1), which might push back this PR out of this release. I'm fine with that - but let me know what you think.

**Comment by aarondr77** [comment]
Okay, the bug above is caused by the fact that the column header is a pure function of the state. So when we set the editor state, the column header automatically opened. It's interesting, that architecture is really helpful most of the time, but as things get more complex, it gets harder to track stuff like that down. If we didn't have consistent naming (like if we changed the names of variables when we created new props) this would be a big issue. Yay for clean code. Anyways, just random thoughts. 

To fix this bug, though, I see three options. 

### Cell Editor 
Add a case to the cell editor to submit the column rename. This would be nice because all of the form handling, stying, etc is already there. 

On the other hand, the cell editor is really just for cell editing, and it would be bad to mix it with column editing, especially when we already have the ColumnHeader component. 

### Column Header component abstraction
Just like we did with the cell editor, we could create a FloatingColumnHeader and use that inside of the grid, and then use the ColumnHeader component in the formula bar. Note: We might want to do some special styling here in the formula bar case because currently, the column header rename input doesn't have any styling.

### Don't handle column headers
Very easily we could just not display column headers in the formula bar. This is reasonable for what a formula bar is, but doesn't make editing long column headers any easier for users.

**Comment by naterush** [comment]
Nice bug investigation! So pretty much:
1. Cell editor opened. This opens column header editor. 
2. Formula bar focused on. 
3. The onBlur for the column header submits it. 

I don't think the ColumnHeaderEditor (whatever it is named) floats currently, not do I think it needs to. So My vote here is we do not refactor to a floating boi. 

I am down to take on the technical debt to make the CellEditor support column header editing when it's in formula bar mode. I think we should just finish this off - again, if this doesn't make the release, I think that's a-ok. Let's take the extra half day and polish this - there's really no rush.

File: mitosheet/src/components/Mito.tsx
Comment (naterush) [inline mitosheet/src/components/Mito.tsx:405]: Remove

File: mitosheet/src/components/endo/FormulaBar.tsx
Comment (naterush) [inline mitosheet/src/components/endo/FormulaBar.tsx:42]: I don't think this is correct to cast to a string like this. We have a function called `getDisplayColumnHeader` - which I think is what you want...

The fact that we _never_ use the `String(` utility on column headers before (but we do have column headers displayed as text all over the place) is good evidence that this approach is not what we want. 

You've gotten got on type casting before. I think when you find yourself converting between types, try and find a similar example to help guide you!

**Review by naterush — CHANGES_REQUESTED** [review]
Two things:
1. Merge in dev (and keep dev's switch.py - ignore this branches changes).
2. We should remove the suggestion box when you're editing a formula. I think this requires an `isColumnHeader` input to the cell editor or something... make sure to test all the paths here thoroughly.

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:402]: Can we move this up to the first condition? E.g. check if rowIndex == -1?

It sounds kinda silly, but I would expect the if statements to flow in the same order that the sheet data does - like column header before everything else. My brain took a second to process as it is - I was expecting column headers first!

File: mitosheet/src/components/endo/FormulaBar.tsx
Comment (naterush) [inline mitosheet/src/components/endo/FormulaBar.tsx:85]: I think we can do the inverse of the check above? Just `props.editorState?.editorLocation !== 'formula bar'` - rather than this more complicated check?

File: mitosheet/src/components/endo/FormulaBar.tsx
Comment (naterush) [inline mitosheet/src/components/endo/FormulaBar.tsx:41]: Per my other comment, doing this if (column header) check as the first condition is much more natural!

File: mitosheet/src/components/endo/ColumnHeader.tsx
Comment (naterush) [inline mitosheet/src/components/endo/ColumnHeader.tsx:345]: I think you can call submit function directly here, as you never need to pass an e value. Wrapping it might just obfuscate things...

Comment (naterush) [reply mitosheet/src/components/endo/ColumnHeader.tsx:345]: It's a bit more verbose, but also more clear I think!

**Review by naterush — CHANGES_REQUESTED** [review]
A few more comments, mostly minor

**Comment by aarondr77** [comment]
The only thing I didn't do was remove the suggestion box while editing a formula. Can you help me understand why you think we should do that? When would the suggestion box be helpful if it wasn't for editing a formula?

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx
Comment (aarondr77) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:384]: remove

**Comment by aarondr77** [comment]
The height of the cell editor on windows and mac is a bit different for some reason, so I updated the height to do what looks best on windows, chrome since that is where most of our users are.

**Comment by naterush** [comment]
@aarondr77 sorry, I meant to say: remove the suggestion box when editing a column header? Can you check if this is good?

**Comment by aarondr77** [comment]
Ooo sweet. Yeah that is taken care of. It just says "You are editing the column header" and doesn't have anything else pop up.

**Pull request merged**
